FrontEnd/src/UserAuthentication/FirestoreDB.js

```diff
@@ -1,16 +1,112 @@
-import {database} from './FirebaseConfig';
-import {doc, setDoc} from 'firebase/firestore';
+import { database } from "./FirebaseConfig";
+import {
+  query,
+  where,
+  collection,
+  doc,
+  getDocs,
+  setDoc,
+  getDoc,
+} from "firebase/firestore";
+import { storage } from "../UserAuthentication/FirebaseConfig";
+import { getDownloadURL, ref, uploadBytes } from "firebase/storage";
+import { v4 } from "uuid";
 
-async function addUser(userCredential, firstName, lastName, userName, signUpEmail){
-    const uid = userCredential.user.uid;
+async function addUser(userId, firstName, lastName, userName, signUpEmail) {
+  const userDocRef = doc(database, "users", userId);
+  return await setDoc(userDocRef, {
+    userID: userId,
+    FirstName: firstName,
+    LastName: lastName,
+    UserName: userName,
+    Email: signUpEmail,
+  });
+}
+
+async function isDatabase() {
+  const usersCollection = query(collection(database, "users"));
+  const querySnapshot = await getDocs(usersCollection);
+  return !querySnapshot.empty;
+}
+
+async function isUsernameUnique(username) {
+  const databaseExists = await isDatabase("users");
+  if (databaseExists === false) {
+    return true;
+  }
+  const usersCollection = collection(database, "users");
+  const check = query(usersCollection, where("UserName", "==", username));
+  const querySnapshot = await getDocs(check);
+  return querySnapshot.empty;
+}
+
+async function updateUserData() {
+  const userDocRef = doc(database, "users", uid);
+  await setDoc(userDocRef, {});
+}
+
+const getUserData = async (uid) => {
+  if (uid === null) {
+    return;
+  }
+  try {
     const userDocRef = doc(database, "users", uid);
-    await setDoc(userDocRef, {
-        uid: uid,
-        FirstName: firstName,
-        LastName: lastName,
-        UserName: userName,
-        Email: signUpEmail,
-    });
+    const userDocSnap = await getDoc(userDocRef);
+
+    if (userDocSnap.exists()) {
+      return userDocSnap.data();
+    }
+  } catch (error) {
+    throw new Error("Error accessing Firestore");
+  }
+};
+
+const uploadPostImage = async (serviceCategory, imageUpload) => {
+  try {
+    const imageRef = ref(
+      storage,
+      `${serviceCategory}/${imageUpload.name + v4()}`
+    );
+    await uploadBytes(imageRef, imageUpload);
+    const downloadURL = await getDownloadURL(imageRef);
+    return downloadURL;
+  } catch (error) {
+    throw new Error("Error uploading image");
+  }
 };
 
-export {addUser};
+const createPost = async (formData, serviceCategory, imageUpload, userId) => {
+  try {
+    const imageURL = await uploadPostImage(serviceCategory, imageUpload);
+    const formDataWithImage = { ...formData, imageURL };
+
+    if (userId === null) {
+      throw new Error("Invalid user ID");
```

Comment:
Let's do this validation in the form, so we don't even try to submit invalid data.
You can keep the "Submit" button disabled until all form inputs are valid.